File: 567.字符串的排列.cpp
/*
 * @lc app=leetcode.cn id=567 lang=cpp
 *
 * [567] 字符串的排列
 */
#include<iostream>
#include<vector>
#include<algorithm>
#include<limits.h>
#include<string>
#include<unordered_map>
using namespace std;
// @lc code=start
class Solution {
public:
    bool checkInclusion(string s1, string s2) {
        unordered_map<char, int> need, window;
        for (char c : s1)
            need[c]++;
        int left = 0, right = 0, validNum = 0; // validNum 记录window内包含的t中的字符(种类）数量
        int lenSubstr = INT_MAX, startSubstr = 0;
        while (right < s2.length()) {
            char cur = s2[right];
            right++;
            if (need.count(cur)) {
                window[cur]++;
                if (window[cur] == need[cur])
                    validNum++;
            }
            while (right - left == s1.length()) {
                if (validNum == need.size()) 
                    return true;
                char shrink = s2[left];
                left++;
                if (need.count(shrink)) {
                    if (need[shrink] == window[shrink])
                        validNum--;
                    window[shrink]--;
                }
            }
        }
        return false;
    }
};
// @lc code=end

File: 567.字符串的排列.java

/*
 * @lc app=leetcode.cn id=567 lang=java
 *
 * [567] 字符串的排列
 */
import java.util.*;

// @lc code=start
class Solution {
    public boolean checkInclusion(String s1, String s2) {
        HashMap<Character, Integer> need = new HashMap<>(), window = new HashMap<>();
        for (int i = 0; i < s1.length(); i++) {
            char c = s1.charAt(i);
            need.put(c, need.getOrDefault(c, 0) + 1);
        }
        int left = 0, right = 0, validNum = 0;
        while (right < s2.length()) {
            char cur = s2.charAt(right);
            right++;
            if (need.containsKey(cur)) {
                window.put(cur, window.getOrDefault(cur, 0) + 1);
                if (need.get(cur).equals(window.get(cur)))
                    validNum++;
                if (validNum == need.size())
                    return true;
            }
            while (right - left >= s1.length()) {
                char shrink = s2.charAt(left);
                if (need.containsKey(shrink)) {
                    // Use equals method when comparing two Integer objects
                    // Only return right answer for -128~127 when using "==" with Integer cache
                    if (window.get(shrink).equals(need.get(shrink))) {
                        validNum--;
                    }
                    window.put(shrink, window.get(shrink) - 1);
                }
                left++;
            }
        }
        return false;
    }
}
// @lc code=end
